feat(ai): add a new app with llm models to track llm runs, store prompts, etc

quevon24 opened the pull request:
## Summary
<!-- What does this fix, how did you fix it, what approach did you take, what gotchas are there in your code or compromises did you make? -->
This PR introduces a new application named `ai` to manage all logic related to Large Language Model (LLM) configuration, execution, and auditing.

This app centralizes the following capabilities:

- Model Storage: It hosts all LLM-related models (Prompt, LLMConfig, LLMPromptSet, LLMRun), separating configuration and audit data from core business logic.
- Utility Layer: It serves as the primary location for helpers and utilities needed to execute and audit tasks efficiently.
- Clean Architecture: This separation ensures the application remains modular, allowing for independent updates to LLM strategy or configuration without impacting core features.

## Deployment

**This PR should:**
<!-- The following labels control the deployment of this PR if they’re applied. -->
<!-- Please put an "X" in the box on ones that apply. -->
<!-- For more details on what pods are affected by each label, see the wiki -->
<!-- https://github.com/freelawproject/courtlistener/wiki/Pull-requests-%60skip%E2%80%90%7Btype%7D%E2%80%90deploy%60-labels -->

<!-- Check here if the entire deployment can be skipped -->
<!-- This might be the case for a small fix, a tweak to documentation or something like that. -->
- [ ] `skip-deploy` (skips everything below)
    <!-- Check here if the web tier can be skipped -->
    <!-- This is the case if you're working on code that doesn't affect the front end, like management commands, tasks, or documentation. -->
    - [ ] `skip-web-deploy`
    <!-- Check here if the deployment to celery can be skipped -->
    <!--This is the case if you make no changes to tasks.py or the code that tasks rely on. -->
    - [ ] `skip-celery-deploy`
    <!-- check this if deployment to cron jobs can be skipped -->
    <!-- This is the case if no changes are made that affect cronjobs. -->
    - [ ] `skip-cronjob-deploy`
    <!-- Deployment of daemons can be skipped -->
    <!-- This is the case if you haven't updated daemons or the code they depend on. -->
    - [ ] `skip-daemon-deploy`

<!-- **If deployment is required:** -->
<!-- What extra steps are needed to deploy this beyond the standard deploy? -->
<!-- Do scripts need to be run or things like that? -->
<!-- If this is more than a quick thing, a new issue should be created in our infra repo: https://github.com/freelawproject/infrastructure/issues/new (if you don’t have access to it, just put the steps here) -->
<!-- Please use an ordered list or delete this if no special steps are required: -->


<!-- Thank you for contributing and filling out this form! -->

quevon24 commented:
Thanks @grossir. I made a few changes in the models. I implemented the admin models. And a second migration to create the configuration for `classify_case_name_by_llm` task.

Once you've applied the migrations, you'll be able to go to the admin panel to view data in: LLMConfig, Prompts, and LLMPromptSet

for LLMRun, it will look something like this:

<img width="2399" height="978" alt="image" src="https://github.com/user-attachments/assets/d0b30937-861a-4750-ad21-0d303d7d7eca" />

<img width="2399" height="978" alt="image" src="https://github.com/user-attachments/assets/073481d5-f95d-461e-9621-1f9c25bf95b7" />

quevon24 commented on cl/ai/models.py:
```diff
@@ -0,0 +1,194 @@
+        return f"{self.provider}/{self.model_name}"
+
+
+class LLMPromptSet(AbstractDateTimeModel):
```
I definitely hear the concern about the number of models, it looks like a lot of glue for what seems like a simple API call. The reason I went with this structure was to decouple the AI's behavior from the application code to avoid a few specific bottlenecks:

- **Versioning & Hot-fixes:** By having these models, we can tweak a prompt in the Admin to fix a hallucination or edge case and have it live in seconds. If we hardcode these, a 2-minute text change requires a full PR and deployment cycle.
- **Model Independence:** LLMConfig lets us swap the engine (e.g., GPT-4o to Gemini) or tweak parameters like temperature without touching business logic.
- **Traceability:** LLMRun acts as our audit log to monitor raw responses for debugging.

Can we lean it down? If the current structure feels over-engineered, we can definitely simplify it:

1. Remove LLMPromptSet: If we're confident we only ever need a 'System' and 'User' prompt pair, we can move those text fields directly onto the LLMTask model and get rid of the PromptSet and Prompt models entirely.
2. Flatten LLMTask: We could manage versions by adding a version field and an is_active boolean directly to the LLMTask. In the code, we’d just fetch the latest active record for a given name (e.g., 'recap_into_opinions_extract_casename').

What do you think? I'm all ears for your thoughts on which direction we should take

grossir commented on cl/ai/models.py:
```diff
@@ -0,0 +1,331 @@
+    llm_key = models.CharField(
+        max_length=255,
+        help_text="The LLM key for this task",
+    )
```
I am not sure what this is referring to. Can you improve the attribute name and the help_text?

flooie replied:
when we push a batch, we generate a key that is used with the batch return.  send 100 opinions in a batch, we need to be able to identify which response goes to which file.

quevon24 replied:
Improving the help_text, as Gianfranco suggests, is not a bad idea.

grossir commented on cl/ai/models.py:
```diff
@@ -0,0 +1,331 @@
+    response_file = models.FileField(
+        help_text="Full response from the LLM provider stored as a file",
+        upload_to=make_response_file_path,
+        storage=IncrementingAWSMediaStorage(),
```
I think this is a public bucket. Do we want the raw response to be in a public bucket?

flooie replied:
no probably not

grossir commented on cl/ai/migrations/0001_initial.sql:
```diff
@@ -0,0 +1,31 @@
+ALTER TABLE "ai_llmtask" ADD CONSTRAINT "ai_llmtask_request_id_80ee2b0f_fk_ai_llmrequest_id" FOREIGN KEY ("request_id") REFERENCES "ai_llmrequest" ("id") DEFERRABLE INITIALLY DEFERRED;
+CREATE INDEX "ai_llmtask_date_created_1c75ad5f" ON "ai_llmtask" ("date_created");
+CREATE INDEX "ai_llmtask_date_modified_d8083f72" ON "ai_llmtask" ("date_modified");
+CREATE INDEX "ai_llmtask_content_type_id_8d14a327" ON "ai_llmtask" ("content_type_id");
```
we could delete this index, since it's contained in the index in line 30

grossir reviewed (commented):
Looks good, mostly storage questions

flooie commented on cl/ai/models.py:
```diff
@@ -0,0 +1,331 @@
+    IN_PROGRESS = 1
+    SUCCEEDED = 2
+    FAILED = 3
+    FINISHED = 4
```
I think SUCCEEDED as it successfully returned a response we want from an LLM call.  Finished means it was processed.  so if we successfully got a good response for case name extraction, until its processed it's not finished.  it will allow us to handle failures in the last step of the process.

quevon24 commented on cl/ai/models.py:
```diff
@@ -0,0 +1,331 @@
+        storage=IncrementingAWSMediaStorage(),
+        max_length=1000,
+        blank=True,
+    )
```
I understand that it is possible, but the `input_file` bucket would have to point to the same bucket where the scraped document is located (otherwise it would generate a non-existent url)

Something like this:

```python
obj = LLMTask()
obj.input_file.name = 'path-to-scrape/file.pdf' # S3 object key
obj.save() 
```

quevon24 commented on cl/ai/models.py:
```diff
@@ -0,0 +1,331 @@
+    batch_response_file = models.FileField(
+        help_text="The batch response file from the LLM provider (e.g., a JSONL file)",
+        upload_to=make_response_file_path,
+        storage=IncrementingAWSMediaStorage(),
```
So we need to implement a new `S3Storage` in `storage.py`

We will use the bucket specified here: `settings.AWS_PRIVATE_STORAGE_BUCKET_NAME`? or would it be a different bucket?

quevon24 commented:
The last few commits implement the Google Gemini batch API integration. Added the wrapper class with support for caching and batch operations, plus a response validator that handles both batch and single API responses. Also refactored the models to use `IntegerChoices` and moved some helpers to a shared location. Tests are all passing now.

grossir commented on cl/ai/models.py:
```diff
@@ -0,0 +1,291 @@
+    llm_key = models.CharField(
+        max_length=255,
+        help_text="A unique identifier for this task, used to map results back from a batch job.",
+    )
```
Maybe this is better, more explicit

"A unique identifier from the LLM  provider system that identifies this this task, used to map results back from a batch job."

grossir commented on cl/ai/models.py:
```diff
@@ -0,0 +1,291 @@
+    make_llm_task_response_file_path,
+)
+from cl.lib.models import AbstractDateTimeModel
+from cl.lib.storage import IncrementingAWSMediaStorage
```
You probably need to implement a new class in `cl.lib.storage` 

Something like this is all we need for it to be private. We don't have to create anything manually on S3, just use the private bucket

```python
class S3LLMResponseStorage(S3Storage):
    """Uses S3 GlacierInstantRetrieval storage class with private ACL"""

    default_acl = "private"
    bucket_name = settings.AWS_PRIVATE_STORAGE_BUCKET_NAME
    file_overwrite = False
```

I see you have created path makers in `cl.lib.model_helpers`. However, I think 

With @flooie  we have been discussing this, and we agree about using llm_responses

The LLMRequest id is unique, and holds a link to all the metadata about this request, so probably the path template should be the "filename"

quevon24 commented on cl/ai/tests.py:
```diff
@@ -0,0 +1,51 @@
+            task=LLMTaskChoices.CASENAME,
+            content_object=self.docket,
+            llm_key="test-key-1",
+        )
```
Good catch on the instance.pk concern. I tested it and confirmed: the current upload_to helpers using `f"{instance.pk}"` only work correctly when setting an existing file path directly `(e.g., task.input_file.name = "some/existing/key.pdf")`, which is the pattern we use in send_gemini_batches.py (See: https://github.com/freelawproject/courtlistener/pull/6907). In that flow, upload_to is never called, so instance.pk is irrelevant.

However, if we ever upload a file directly via FileField.save(), Django calls upload_to before the instance is saved to the DB, so instance.pk is None and the path becomes something like `llm-tasks/2026/02/12/None.pdf`.

I can switch the path helpers from instance.pk to `uuid4().hex` so filenames are always unique regardless of save order. For example: `llm-tasks/2026/02/12/a3b2c1d4e5f6789012345678abcdef01.pdf`. That way we're safe for both use cases (existing paths and direct uploads). What do you think?

Regarding `input_file` taking an existing public bucket path: yes, it works. Django's FileField is just a varchar column. Setting `task.input_file.name = "some/existing/path.pdf"` and saving simply stores that string in the DB with no validation that the file exists at that path. It doesn't trigger `upload_to` or any storage check.

quevon24 commented:
Can you take a look at it, @mlissner?

chaco-fl commented:
Customers RDS Updated 
<img width="718" height="576" alt="Screenshot 2026-02-27 at 11 27 25" src="https://github.com/user-attachments/assets/d5fe1b93-19a0-49cc-8b98-d97d7caad685" />


Adding the new table to each replication.
